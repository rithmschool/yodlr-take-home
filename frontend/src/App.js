import React, { useEffect } from 'react';
import './App.css';
import { useDispatch } from 'react-redux';
import { loadUsers } from './reducer/users/action';
import Nav from './Nav';
import Routes from './Routes';


function App() {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(loadUsers());
  }, [dispatch]);

  return (
    <div className="App">
      <Nav />
      <div className="container">
        <Routes />
      </div>
    </div>
  );
}

export default App;
