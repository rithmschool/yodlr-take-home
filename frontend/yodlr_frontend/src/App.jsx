import React from "react";
import { Routes, Route } from "react-router-dom";
import { Container } from "react-bootstrap";
import Home from "../components/Home";
import Admin from "../components/Admin";
import Register from "../components/Register";
import Navbar from "../components/Navbar";


function App() {
  return (
    <div className="App">
      <Navbar />
        <Container className="mb-4">
          <Routes>
            <Route path="/Home" element={<Home />} />
            <Route path="/Admin" element={<Admin />} />
            <Route path="/Register" element={<Register />} />
          </Routes>
        </Container>
    </div>
  )
}

export default App;
