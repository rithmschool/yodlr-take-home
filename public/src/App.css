.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}
.card {
  border: 1px solid grey;
  border-radius: 8px;
  margin-bottom: 1em;
  width: 50%;
  display: inline-block;
}

.card button {
  background-color: purple;
  color: white;
  border-radius: 4px;
}
@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
