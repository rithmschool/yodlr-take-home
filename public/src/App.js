import React, { useState, useEffect } from 'react';
import  { CurrentUserContext, UsersContext } from './YodlrContext';
import Routes from './Routes';
import YodlrApi from './api.js';
import './App.css';

function App() {
  // create states for needed information
  const [isLoading, setIsLoading] = useState(true);
  const [currentUser, setCurrentUser] = useState(null);
  const [users, setUsers] = useState(null);
  

  // get needed information on load if user is stored in localstorage
  useEffect(() => {
    async function onLoad() {
      try{
      const id = localStorage.getItem('id');
      const token = localStorage.getItem('token');
      // if (token && id) {YodlrApi.token = token;
      // } else {throw new Error()}
      YodlrApi.token = token;
      const user = await YodlrApi.getUser(id);
      setCurrentUser({token: token, firstName: user.firstName, lastName: user.lastName, email: user.email});
      const newUsers = await YodlrApi.getUsers();
      console.log(newUsers)
      setUsers(newUsers);
      setIsLoading(false)
      } catch (err) {
        setCurrentUser(false);
      }
    }
    onLoad();
  }, [isLoading]);

  // login user
  async function login(formData) {
    const res = await YodlrApi.login(formData);
    setCurrentUser({token: res, id: formData.id});
    localStorage.setItem('token', res);
    localStorage.setItem('id', formData.id);
    setIsLoading(true);
  }

  // logout user
  async function logout() {
    await YodlrApi.logout();
    setCurrentUser({});
    localStorage.clear();
    setIsLoading(true);
  }

  // register new user
  async function signup (formData) {
    await YodlrApi.signup(formData);
    return (<div><h3>Thank you for registering!</h3></div>);
  }

  return (
    <div className="App">
      <CurrentUserContext.Provider value={currentUser}>
      <UsersContext.Provider value={users}>
        <Routes login={login} logout={logout} signup={signup} />
      </UsersContext.Provider>
      </CurrentUserContext.Provider>
    </div>
  );
}

export default App;
