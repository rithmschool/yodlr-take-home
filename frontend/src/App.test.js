import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react';
import App from './App';
import { useSelector, useDispatch } from 'react-redux';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}));

beforeEach(function() {
  useSelector.mockImplementation((selectorFn) =>
    selectorFn({
      users: {
        0: {
          id: 0, firstName: "test", lastName: "user", state: "active"
        },
        1: {
          id: 1, firstName: "test1", lastName: "user", state: "pending"
        }
      }
    })
  );
})

// smoke test
it("renders without crashing", function() {
  render(<MemoryRouter><App /></MemoryRouter>);
});

// snapshot test
it("matches snapshot", function() {
  const { asFragment } = render(<MemoryRouter><App /></MemoryRouter>);
  expect(asFragment()).toMatchSnapshot();
});


describe("NavBar links test", function() {
  it("goes to /signup", async function() {
    const { getByText, findByText } = render(<MemoryRouter><App /></MemoryRouter>);
    const signupLink = await findByText("Sign Up");

    fireEvent.click(signupLink);
    expect(getByText("First Name")).toBeInTheDocument();
    expect(getByText("Last Name")).toBeInTheDocument();
    expect(getByText("Email")).toBeInTheDocument();
    expect(getByText("Sign up")).toBeInTheDocument();
  });

  it("goes to /admin", async function() {
    const { getByText, findByText } = render(<MemoryRouter><App /></MemoryRouter>);
    const adminLink = await findByText("Admin");

    fireEvent.click(adminLink);
    expect(getByText("First Name")).toBeInTheDocument();
    expect(getByText("Last Name")).toBeInTheDocument();
    expect(getByText("Email")).toBeInTheDocument();
    expect(getByText("State")).toBeInTheDocument();
  });

  it("goes to /", async function() {
    const { getByText, findByText } = render(<MemoryRouter><App /></MemoryRouter>);
    const adminLink = await findByText("Admin");
    fireEvent.click(adminLink);
    expect(getByText("First Name")).toBeInTheDocument();
    expect(getByText("Last Name")).toBeInTheDocument();
    expect(getByText("Email")).toBeInTheDocument();
    expect(getByText("State")).toBeInTheDocument();

    const homeLink = await findByText("Home");
    fireEvent.click(homeLink);
    expect(getByText("Welcome to Yodlr"));
  });
});