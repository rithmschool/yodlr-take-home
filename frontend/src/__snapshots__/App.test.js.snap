// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="Nav navbar navbar-light bg-white"
    >
      <span
        class="navbar-brand"
      >
        Yodlr
      </span>
      <span>
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          Home
        </a>
        <a
          href="/signup"
        >
          Sign Up
        </a>
        <a
          href="/admin"
        >
          Admin
        </a>
      </span>
    </nav>
    <div
      class="container"
    >
      <div
        class="row justify-content-center mt-5"
      >
        <div
          class="card col-sm-8"
        >
          <div
            class="card-body"
          >
            <h3>
              Welcome to Yodlr
            </h3>
            <h5>
              <i>
                Please feel free to explore and learn more about what we can offer!
              </i>
            </h5>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
